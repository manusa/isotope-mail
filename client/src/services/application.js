import {setUserCredentials} from '../actions/application';

export function login(dispatch, credentials) {
  fetch('http://localhost:9010/v1/application/login', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(credentials)
  })
    .then(response => (response.json()))
    .then(json =>
      dispatch(setUserCredentials(json))
    );
}
