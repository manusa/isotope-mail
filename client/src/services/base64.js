// https://stackoverflow.com/a/30106551/999375
// https://tools.ietf.org/html/rfc4648#page-6 - Standard Base64 Alphabet
// https://tools.ietf.org/html/rfc4648#page-8 - Url Safe  Base64 Alphabet

export function unicodeUrlBtoa(string) {
  return btoa(encodeURIComponent(string).replace(/%([0-9A-F]{2})/g,
    (match, p1) => String.fromCharCode(`0x${p1}`)
  )).replace(/\+/g, '-').replace(/\//g, '_');
}

export function unicodeUrlAtob(string) {
  const replacedString = string.replace(/-/g, '+').replace(/_/g, '/');
  return decodeURIComponent(atob(replacedString)
    .split('')
    // eslint-disable-next-line prefer-template
    .map(c => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2))
    .join(''));
}

