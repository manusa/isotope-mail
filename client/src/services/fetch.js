export const HttpHeaders = {
  ISOTOPE_CREDENTIALS: 'X-Isotope-Credentials',
  ISOTOPE_SALT: 'X-Isotope-Salt'
};

/**
 * Converts response Promise to json Promise or throws an error.
 *
 * @param {Response}response
 * @returns {PromiseLike<object>}
 */
export function toJson(response) {
  if (!response.ok) {
    throw Error(`${response.status}`);
  }
  return response.json();
}

/**
 * Creates a new headers object including Isotope specific credentials headers for server authentication.
 *
 * @param {{}} credentials
 * @param {{}} [originalHeaders={}] originalHeaders
 * @returns {{}}
 */
export function credentialsHeaders(credentials, originalHeaders) {
  const newHeaders = Object.assign({}, originalHeaders ? originalHeaders : {});
  newHeaders[HttpHeaders.ISOTOPE_CREDENTIALS] = credentials.encrypted;
  newHeaders[HttpHeaders.ISOTOPE_SALT] = credentials.salt;
  return newHeaders;
}
