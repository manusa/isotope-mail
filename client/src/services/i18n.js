import i18n from 'i18next';
import XHR from 'i18next-xhr-backend';

const userLanguage = () => navigator.language;

i18n
  .use(XHR)
  .init({
    lng: userLanguage(),
    fallbackLng: 'en',
    ns: ['isotope'],
    defaultNS: 'isotope',
    backend: {
      loadPath: '/assets/locales/{{lng}}/{{ns}}.json'
    },
    interpolation: {
      escapeValue: false
    },
    react: {
      wait: false,
      withRef: false,
      bindI18n: 'languageChanged loaded',
      bindStore: 'added removed',
      nsMode: 'default'
    }
  });

export default i18n;
