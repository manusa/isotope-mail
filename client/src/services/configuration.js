import {URLS} from './url';
import {toJson} from './fetch';

/**
 * @module services/configuration
 */
/**
 * @typedef {Object} Configuration
 * @property {Array} _links - HATEOAS links for application
 */

const TABLET_BREAKPOINT_PX = 600;

/**
 * Fetches configuration from BE
 * @returns {Promise<Configuration>}
 */
export async function fetchConfiguration() {
  const response = await fetch(URLS.CONFIGURATION);
  if (response.ok) {
    return await toJson(response);
  }
  return null;
}

/**
 * Checks if the current viewport is desktop sized
 *
 * @returns {boolean} true if current window size is a valid 'desktop' size, false otherwise
 */
export function isDesktop() {
  return Math.max(document.documentElement.clientWidth, window.innerWidth) > TABLET_BREAKPOINT_PX;
}
