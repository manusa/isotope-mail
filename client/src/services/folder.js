import get from 'lodash/get';
import {
  backendRequest as applicationBackendRequest,
  backendRequestCompleted as applicationBackendRequestCompleted, createFolder,
  renameFolder as renameFolderAction, renameFolderOk
} from '../actions/application';
import {backendRequest, setFolders, updateFolder} from '../actions/folders';
import {renameCache} from '../actions/messages';
import {unicodeUrlAtob as atob, unicodeUrlBtoa as btoa} from './base64';
import {deleteMessageCache, renameMessageCache} from './indexed-db';
import {abortControllerWrappers, abortFetch, credentialsHeaders, refreshCredentials, toJson} from './fetch';
import {notifyNewMail} from './notification';
import {getIsotopeConfiguration} from '../selectors/globals';

export const FolderTypes = Object.freeze({
  INBOX: {serverName: 'INBOX', icon: 'inbox', position: 0},
  DRAFTS: {attribute: '\\Drafts', icon: 'drafts', position: 1},
  SENT: {attribute: '\\Sent', icon: 'send', position: 2},
  TRASH: {attribute: '\\Trash', icon: 'delete', position: 3},
  JUNK: {attribute: '\\Junk', icon: 'report', position: 4},
  FOLDER: {icon: 'folder'}
});


/**
 * Finds the trash folder from within the current folder state's exploded items.
 *
 * TODO: Create cached selector within REDUX state
 * @param foldersState
 * @returns {any}
 */
export const findTrashFolder = foldersState =>
  Object.values(foldersState.explodedItems).find(f => f.type === FolderTypes.TRASH);

/**
 * Returns the id of the provided folder and all its children.
 *
 * @param {object} folder
 * @returns {Array.<string>}
 */
export function gatherFolderIds(folder) {
  const gatheredIds = [];
  const gatherFolderIdsInner = folderInner => {
    gatheredIds.push(folderInner.folderId);
    if (Array.isArray(folderInner.children)) {
      folderInner.children.forEach(gatherFolderIdsInner);
    }
    return gatheredIds;
  };
  return gatherFolderIdsInner(folder);
}

/**
 * Renames message cache for a folder and all its children in Redux store and IndexedDB.
 *
 * This step is necessary as the "key" for the folder's message cache is the folder Id which is
 * a base64 encoded string of the folder URL (which includes the name).
 *
 * @param dispatch
 * @param user
 * @param renamedParentFolder
 * @private
 */
function _renameMessageCache(dispatch, user, renamedParentFolder) {
  if (!renamedParentFolder || !renamedParentFolder.previousFolderId) {
    return;
  }
  const decodedPreviousFolderId = atob(renamedParentFolder.previousFolderId);
  const decodedNewFolderId = atob(renamedParentFolder.folderId);
  // Rename all child folders
  gatherFolderIds(renamedParentFolder)
    .forEach(childFolderId => {
      const decodedChildFolderId = atob(childFolderId);
      const decodedPreviousChildFolderId = decodedChildFolderId.replace(decodedNewFolderId, decodedPreviousFolderId);
      const previousChildFolderId = btoa(decodedPreviousChildFolderId);
      // Rename cache in Redux store
      dispatch(renameCache(previousChildFolderId, childFolderId));
      // Rename cache in IndexedDB database
      renameMessageCache(user.id, user.hash, previousChildFolderId, childFolderId);
    });
}

/**
 * Processes an initial folder list and adds the corresponding {@link FolderTypes}
 *
 * @param initialFolders {Array}
 * @returns {Array}
 */
export function processFolders(initialFolders) {
  if (!initialFolders) {
    return null;
  }
  const folders = [];
  const specialFolders = [];
  initialFolders.map(folder => {
    if (get(folder, 'fullName', '').toUpperCase() === FolderTypes.INBOX.serverName) {
      folder.type = FolderTypes.INBOX;
      specialFolders[folder.type.position] = folder;
    } else {
      folder.type = FolderTypes.FOLDER;
      // Identify special folder
      let special = false;
      for (const t of [FolderTypes.DRAFTS, FolderTypes.SENT, FolderTypes.TRASH, FolderTypes.JUNK]) {
        if (folder.attributes && folder.attributes.includes(t.attribute)) {
          folder.type = t;
          specialFolders[folder.type.position] = folder;
          special = true;
          break;
        }
      }
      // If regular folder just add to the folder array
      if (!special) {
        folders.push(folder);
      }
    }
    folder.children = processFolders(folder.children);
  });
  // Insert special folders in specific positions
  specialFolders.reverse().forEach(f => folders.unshift(f));
  return folders;
}

/**
 * Returns an array with a folder tree with just folderId and children properties.
 *
 * @param {Array} folderTree
 * @returns {{folderId, children: *}[]}
 */
export function removeAttributesFromFolders(folderTree) {
  return [...folderTree].map(({folderId, children}) => ({folderId, children: removeAttributesFromFolders(children)}));
}

/**
 * Builds an object containing all of the folders and their children at the same level from the provided folderTree
 */
export function explodeFolders(folderTree, explodedFolders = {}) {
  folderTree.forEach(f => {
    explodedFolders[f.folderId] = {...f};
    if (f.children && f.children.length > 0) {
      explodeFolders(f.children, explodedFolders);
      explodedFolders[f.folderId].children = removeAttributesFromFolders(explodedFolders[f.folderId].children);
    }
  });
  return explodedFolders;
}

/**
 * Returns the setFolders action with the preprocessed folders in the payload.
 *
 * @param dispatch
 * @param credentials
 * @param loadChildren
 * @returns {Promise<*>}
 */
export async function getFolders(dispatch, credentials, loadChildren) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  abortControllerWrappers.getFoldersAbortController = new AbortController();
  const signal = abortControllerWrappers.getFoldersAbortController.signal;

  const url = new URL(getIsotopeConfiguration()._links.folders.href, window.location.origin);
  if (loadChildren) {
    url.search = new URLSearchParams({loadChildren: true}).toString();
  }
  dispatch(backendRequest());
  const response = await fetch(url, {
    method: 'GET',
    headers: credentialsHeaders(credentials),
    signal: signal
  });
  await refreshCredentials(dispatch, response);
  const folders = await toJson(response);
  const foldersAction = setFolders(folders);
  const inbox = Object.values(foldersAction.payload).find(f => f.type === FolderTypes.INBOX);
  if (inbox && inbox.newMessageCount > 0) {
    notifyNewMail();
  }
  return dispatch(foldersAction);
}

export function renameFolder(dispatch, user, folderToRename, newName) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  fetch(getIsotopeConfiguration()._links['folders.rename'].href.replace('{folderId}', folderToRename.folderId), {
    method: 'PUT',
    headers: credentialsHeaders(user.credentials, {'Content-Type': 'application/json'}),
    body: newName
  })
    .then(toJson)
    .then(renamedFolderParent => {
      if (renamedFolderParent.fullName.length === 0) {
        // Root folder (replace current folder tree)
        dispatch(setFolders(renamedFolderParent.children));
      } else {
        dispatch(updateFolder(renamedFolderParent));
      }
      renamedFolderParent.children
        .filter(f => f.previousFolderId)
        .forEach(f => {
          _renameMessageCache(dispatch, user, f);
          dispatch(renameFolderOk(f.previousFolderId, f.folderId));
        });
      dispatch(renameFolderAction(null));
      dispatch(applicationBackendRequestCompleted());
    })
    .catch(() => {
      dispatch(applicationBackendRequestCompleted());
    });
}

/**
 * Triggers BE API to move the provided folderToMove as a child of the specified targetFolder.
 *
 * REDUX store will only be updated once an OK response is received from te server.
 *
 * @param dispatch
 * @param user
 * @param folderToMove
 * @param targetFolder
 */
export function moveFolder(dispatch, user, folderToMove, targetFolder) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  const targetFolderId = targetFolder ? targetFolder.folderId : null;
  fetch(getIsotopeConfiguration()._links['folders.move'].href.replace('{folderId}', folderToMove.folderId), {
    method: 'PUT',
    headers: credentialsHeaders(user.credentials, {'Content-Type': 'application/json'}),
    body: targetFolderId
  })
    .then(toJson)
    .then(updatedTargetFolder => {
      if (targetFolder) {
        dispatch(updateFolder(updatedTargetFolder));
      } else {
        dispatch(setFolders(updatedTargetFolder.children));
      }
      updatedTargetFolder.children
        .filter(f => f.previousFolderId)
        .forEach(f => {
          _renameMessageCache(dispatch, user, f);
          dispatch(renameFolderOk(f.previousFolderId, f.folderId));
        });
      dispatch(applicationBackendRequestCompleted());
    })
    .catch(() => {
      dispatch(applicationBackendRequestCompleted());
    });
}

/**
 * Triggers BE API to premanently delete the provided folderToDelete and its children.
 *
 * REDUX store will only be updated once an OK response is received from te server.
 *
 * @param dispatch
 * @param user
 * @param folderToDelete
 */
export function deleteFolder(dispatch, user, folderToDelete) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  fetch(getIsotopeConfiguration()._links['folders.delete'].href.replace('{folderId}', folderToDelete.folderId), {
    method: 'DELETE',
    headers: credentialsHeaders(user.credentials)
  })
    .then(toJson)
    .then(updatedParentFolder => {
      dispatch(updateFolder(updatedParentFolder));
      const folderIds = gatherFolderIds(folderToDelete);
      deleteMessageCache(user.id, user.hash, folderIds);
      dispatch(applicationBackendRequestCompleted());
    })
    .catch(() => {
      dispatch(applicationBackendRequestCompleted());
    });
}

/**
 * Creates a new 1st/root level folder.
 *
 * @param dispatch {Dispatch & function}
 * @param user
 * @param newFolderName name for the folder to be created
 */
export function createRootFolder(dispatch, user, newFolderName) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  fetch(getIsotopeConfiguration()._links.folders.href, {
    method: 'POST',
    headers: credentialsHeaders(user.credentials),
    body: newFolderName
  })
    .then(toJson)
    .then(folders => {
      dispatch(setFolders(folders));
      dispatch(createFolder(null));
      dispatch(applicationBackendRequestCompleted());
    })
    .catch(() => {
      dispatch(applicationBackendRequestCompleted());
    });
}


/**
 * Creates a new folder under the specified parentFolder.
 *
 * @param dispatch {Dispatch & function}
 * @param user
 * @param parentFolder {object} folder to add new child folder
 * @param newFolderName name for the folder to be created
 */
export function createChildFolder(dispatch, user, parentFolder, newFolderName) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  fetch(getIsotopeConfiguration()._links['folders.self'].href.replace('{folderId}', parentFolder.folderId), {
    method: 'POST',
    headers: credentialsHeaders(user.credentials),
    body: newFolderName
  })
    .then(toJson)
    .then(updatedParentFolder => {
      dispatch(updateFolder(updatedParentFolder));
      dispatch(createFolder(null));
      dispatch(applicationBackendRequestCompleted());
    })
    .catch(() => {
      dispatch(applicationBackendRequestCompleted());
    });
}
