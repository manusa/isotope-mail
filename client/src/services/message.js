import {
  backendRequest as applicationBackendRequest,
  backendRequestCompleted as applicationBackendRequestCompleted,
  preDownloadMessages
} from '../actions/application';
import {
  backendRequest,
  backendRequestCompleted,
  deleteFromCache, lockMessages, setFolderCache,
  setSelected, unlockMessages,
  updateCache, updateCacheIfExist
} from '../actions/messages';
import {updateFolder} from '../actions/folders';
import {abortControllerWrappers, abortFetch, AuthenticationException, credentialsHeaders, toJson} from './fetch';
import {persistMessageCache} from './indexed-db';
import {notifyNewMail} from './notification';
import {FolderTypes} from './folder';

const _eventSourceWrappers = {};

function _closeEventSource(dispatch, es) {
  if (es && es.close) {
    es.close();
    if (!es.DISPATCHED) {
      dispatch(backendRequestCompleted());
      es.DISPATCHED = true;
      es.resolvePromise();
    }
  }
}

/**
 * "Aborts" the event source (SSE received stream) to reset folder message cache.
 *
 * This will cancel current message poll.
 *
 * @param dispatch
 */
export function closeResetFolderMessagesCacheEventSource(dispatch) {
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
}

/**
 *
 * @param dispatch
 * @param user
 * @param folder
 */
export async function resetFolderMessagesCache(dispatch, user, folder) {
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  if (folder && folder._links) {
    const allMessages = [];
    // Prefer EventSourcePolyfill instead of EventSource to allow sending HTTP headers in all browsers
    const es = new window.EventSourcePolyfill(folder._links.messages.href,
      {
        headers: credentialsHeaders(user.credentials)
      });
    _eventSourceWrappers.resetFolderMessagesCache = es;
    dispatch(backendRequest());
    let isFirstBatch = true;
    es.onmessage = e => {
      const messages = JSON.parse(e.data);
      allMessages.push(...messages);

      // Remove UIDs not included in batch update from store
      const originalUids = messages.map(m => m.uid);
      let maxUid = null;
      let minUid = null;
      if (originalUids.length > 0) {
        maxUid = originalUids.reduce((a, b) => Math.max(a, b));
        minUid = originalUids.reduce((a, b) => Math.min(a, b));
        const completeUidSequence = Array(maxUid - minUid + 1).fill('').map((v, i) => i + minUid);
        const uidsToRemove = completeUidSequence.filter(uid => !originalUids.includes(uid));
        if (uidsToRemove.length > 0) {
          dispatch(deleteFromCache(folder, uidsToRemove.map(uid => ({uid}))));
        }
      }

      dispatch(updateCache(folder, messages));

      // This is the First Batch -> Delete all messages with a higher UID not included in the batch
      if (isFirstBatch) {
        isFirstBatch = false;
        if (maxUid) {
          dispatch(deleteFromCache(folder, [], {from: maxUid + 1}));
        } else {
          // Clear empty folder if first batch has no uids
          dispatch(deleteFromCache(folder, [], {from: 0}));
        }
      }

      // This is the last batch -> persistMessageCache with all new gathered messages
      if (e.lastEventId === '1') {
        _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
        // Remove uids not included in batch (0-minUidInBatch)
        if (minUid) {
          dispatch(deleteFromCache(folder, [], {to: minUid - 1}));
        }
        // Manually persist newest version of message cache
        persistMessageCache(user.id, user.hash, folder, [...allMessages]);
      }
    };
    // Return a promise
    return new Promise((resolve, reject) => {
      // Store resolve function that will be called in _closeEventSource
      es.resolvePromise = resolve;
      es.onerror = ({status, message}) => {
        if (status === 401) {
          reject(new AuthenticationException(message));
        }
      };
    });
  }
  return null;
}

export function preloadMessages(dispatch, credentials, folder, messageUids) {
  if (messageUids.length === 0) {
    return;
  }
  const url = new URL(folder._links.messages.href);
  messageUids.forEach(messageUid => url.searchParams.append('id', messageUid));
  fetch(url, {
    method: 'GET',
    headers: credentialsHeaders(credentials)
  })
    .then(toJson)
    .then(messages => {
      if (folder.type === FolderTypes.INBOX && messages.some(m => m.recent === true)) {
        notifyNewMail();
      }
      dispatch(preDownloadMessages(messages));
    })
    .catch(() => {/* Ignore any preload error */});
}

export function downloadAttachment(credentials, attachment) {
  const fetch$ = fetch(attachment._links.download.href, {
    method: 'GET',
    headers: credentialsHeaders(credentials)
  }).then(response => response.blob());
  fetch$
    .then(blob => {
      if (navigator.msSaveBlob) {
        navigator.msSaveBlob(blob, attachment.fileName);
      } else {
        const url = URL.createObjectURL(blob);
        const tempLink = document.createElement('a');
        tempLink.href = url;
        tempLink.download = attachment.fileName;
        document.body.appendChild(tempLink);
        // tempLink.click();
        tempLink.dispatchEvent(new MouseEvent('click', {bubbles: true, cancelable: true, view: window}));
        document.body.removeChild(tempLink);
        URL.revokeObjectURL(url);
      }
    });
  return fetch$;
}

/**
 * Moves the provided array of messages from the original fromFolder to the target toFolder.
 *
 * This method is optimistic and will calculate fromFolder message counts before any server response is received.
 *
 * @param dispatch {!function}
 * @param credentials {!Object}
 * @param fromFolder {!Object}
 * @param toFolder {!Object}
 * @param messages {!Array.<Object>}
 */
export function moveMessages(dispatch, credentials, fromFolder, toFolder, messages) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  const fromFolderUpdated = {
    ...fromFolder
  };
  messages.forEach(m => {
    dispatch(setSelected([m], false));
    // Calculate new fromFolder values
    fromFolderUpdated.messageCount -= 1;
    fromFolderUpdated.unreadMessageCount -= (m.seen ? 0 : 1);
    fromFolderUpdated.newMessageCount -= (m.recent ? 1 : 0);
  });

  // Update state with expected response from server
  dispatch(deleteFromCache(fromFolder, messages));
  dispatch(updateFolder(fromFolderUpdated));
  dispatch(lockMessages(messages));
  fetch(fromFolder._links['message.move.bulk'].href.replace('{toFolderId}', toFolder.folderId), {
    method: 'PUT',
    headers: credentialsHeaders(credentials, {'Content-Type': 'application/json'}),
    body: JSON.stringify(messages.map(m => m.uid))
  })
    .then(toJson)
    .then(newMessages => {
      dispatch(unlockMessages(messages));
      if (Array.isArray(newMessages)) {
        dispatch(updateCache(toFolder, newMessages));
        // Update folder info with the last message (contains the most recent information)
        dispatch(updateFolder(newMessages[newMessages.length - 1].folder));
      }
    })
    .catch(() => {
      dispatch(unlockMessages(messages));
      // Rollback state from dispatched expected responses
      dispatch(updateCache(fromFolder, messages));
      dispatch(updateFolder(fromFolder));
    });
}

export function setMessagesSeen(dispatch, credentials, folder, messages, seen) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  const messagesToUpdate = [];
  // Update state with expected response from server
  const expectedUpdatedFolder = {...folder};
  messages.forEach(m => {
    // Calculate new folder values
    if (m.seen !== seen) {
      messagesToUpdate.push({...m, seen: seen});
      expectedUpdatedFolder.unreadMessageCount += seen ? -1 : +1;
    }
  });
  dispatch(updateCacheIfExist(folder, messagesToUpdate, true));
  dispatch(updateFolder(expectedUpdatedFolder));
  dispatch(lockMessages(messagesToUpdate));

  fetch(folder._links['message.seen.bulk'].href.replace('{seen}', seen.toString()), {
    method: 'PUT',
    headers: credentialsHeaders(credentials, {'Content-Type': 'application/json'}),
    body: JSON.stringify(messagesToUpdate.map(m => m.uid))
  })
    .then(response => {
      dispatch(unlockMessages(messagesToUpdate));
      if (!response.ok) {
        // Rollback state from dispatched expected responses
        dispatch(updateCacheIfExist(folder, messages));
        dispatch(updateFolder(folder));
      }
    });
}

export function setMessageFlagged(dispatch, credentials, folder, message, flagged) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  dispatch(updateCacheIfExist(folder, [{...message, flagged}], true));
  dispatch(lockMessages([message]));
  fetch(folder._links['message.flagged'].href.replace('{messageId}', message.uid), {
    method: 'PUT',
    headers: credentialsHeaders(credentials, {'Content-Type': 'application/json'}),
    body: JSON.stringify(flagged)
  })
    .then(response => {
      dispatch(unlockMessages([message]));
      if (!response.ok) {
        // Rollback state from dispatched expected responses
        dispatch(updateCacheIfExist(folder, [message]));
      }
    });
}

export async function deleteAllFolderMessages(dispatch, credentials, folder) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  try {
    const updatedFolder = await fetch(folder._links.messages.href, {
      method: 'DELETE',
      headers: credentialsHeaders(credentials)
    }).then(toJson);
    dispatch(updateFolder(updatedFolder));
    dispatch(setFolderCache(updatedFolder, []));
    dispatch(applicationBackendRequestCompleted());
  } catch (e) {
    dispatch(applicationBackendRequestCompleted());
  }
}

export function deleteMessages(dispatch, credentials, folder, messages) {
  if (messages.length === 0) {
    return;
  }
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  const url = new URL(folder._links.messages.href);
  const messagesToDelete = [];
  // Update state with expected response from server
  const expectedUpdatedFolder = {...folder};
  messages.forEach(message => {
    url.searchParams.append('id', message.uid);
    messagesToDelete.push({...message, deleted: true});
    expectedUpdatedFolder.deletedMessageCount++;
    expectedUpdatedFolder.unreadMessageCount = message.seen ? expectedUpdatedFolder.unreadMessageCount
      : expectedUpdatedFolder.unreadMessageCount - 1;
    expectedUpdatedFolder.newMessageCount = message.recent ? expectedUpdatedFolder.newMessageCount - 1
      : expectedUpdatedFolder.newMessageCount;
  });
  dispatch(updateCacheIfExist(folder, messagesToDelete, true));
  dispatch(setSelected(messagesToDelete, false));
  dispatch(updateFolder(expectedUpdatedFolder));
  dispatch(lockMessages(messagesToDelete));

  fetch(url, {
    method: 'DELETE',
    headers: credentialsHeaders(credentials)
  })
    .then(toJson)
    .then(updatedFolder => {
      dispatch(unlockMessages(messagesToDelete));
      dispatch(updateFolder(updatedFolder));
      dispatch(deleteFromCache(updatedFolder, messagesToDelete));
    })
    .catch(() => {
      dispatch(unlockMessages(messagesToDelete));
      // Rollback state from dispatched expected responses
      dispatch(updateCacheIfExist(folder, messages));
      dispatch(updateFolder(folder));
    });
}
