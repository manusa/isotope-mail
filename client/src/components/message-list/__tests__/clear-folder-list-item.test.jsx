import React from 'react';
import {shallow} from 'enzyme/build/index';
import {MOCK_STORE} from '../../../__testutils__/store';
import ConnectedClearFolderListItem, {ClearFolderListItem} from '../clear-folder-list-item';
import {FolderTypes} from '../../../services/folder';
import * as messageService from '../../../services/message';

describe('ClearFolderListItem component test suite', () => {
  describe('Snapshot render', () => {
    test('Selected folder NOT trash nor junk, should not render', () => {
      // Given
      const props = {
        selectedFolder: {}
      };
      // When
      const clearFolderListItem = shallow(<ClearFolderListItem {...props} />);
      // Then
      expect(clearFolderListItem).toMatchSnapshot();
    });
    test('Selected folder is trash, should render', () => {
      // Given
      const props = {
        t: key => key,
        selectedFolder: {type: FolderTypes.TRASH}
      };
      // When
      const clearFolderListItem = shallow(<ClearFolderListItem {...props} />);
      // Then
      expect(clearFolderListItem).toMatchSnapshot();
    });
  });
  test('Selected folder is junk, should render', () => {
    // Given
    const props = {
      t: key => key,
      selectedFolder: {type: FolderTypes.JUNK}
    };
    // When
    const clearFolderListItem = shallow(<ClearFolderListItem {...props} />);
    // Then
    expect(clearFolderListItem).toMatchSnapshot();
  });
  describe('nested component functions', () => {
    test('Button.onclick, should set dialog visible', () => {
      // Given
      const props = {
        t: key => key,
        selectedFolder: {type: FolderTypes.TRASH}
      };
      const clearFolderListItem = shallow(<ClearFolderListItem {...props} />);
      // When
      clearFolderListItem.find('Button').props().onClick();
      // Then
      expect(clearFolderListItem.find('Translate(ConfirmClearFolderDialog)').props().visible).toBe(true);
    });
    test('ConfirmClearFolderDialog.cancelAction, should set dialog not visible', () => {
      // Given
      const props = {
        t: key => key,
        selectedFolder: {type: FolderTypes.TRASH}
      };
      const clearFolderListItem = shallow(<ClearFolderListItem {...props} />);
      clearFolderListItem.find('Button').props().onClick();
      // When
      clearFolderListItem.find('Translate(ConfirmClearFolderDialog)').props().cancelAction();
      // Then
      expect(clearFolderListItem.find('Translate(ConfirmClearFolderDialog)').props().visible).toBe(false);
    });
    test('ConfirmClearFolderDialog.clearFolder, should trigger function', () => {
      // Given
      const props = {
        t: key => key,
        selectedFolder: {type: FolderTypes.TRASH},
        clearFolder: jest.fn()
      };
      const clearFolderListItem = shallow(<ClearFolderListItem {...props} />);
      // When
      clearFolderListItem.find('Translate(ConfirmClearFolderDialog)').props().deleteAction();
      // Then
      expect(props.clearFolder).toHaveBeenCalledTimes(1);
    });
  });
  describe('Connect functions', () => {
    test('clearFolder, dispatch actions triggered', () => {
      const props = {
        store: MOCK_STORE,
        selectedFolder: {}
      };
      messageService.deleteAllFolderMessages = jest.fn();
      const clearFolderListItem = shallow(<ConnectedClearFolderListItem {...props} />);
      // When
      clearFolderListItem.props().clearFolder();
      // Then
      expect(messageService.deleteAllFolderMessages).toHaveBeenCalledTimes(1);
    });
  });
});
