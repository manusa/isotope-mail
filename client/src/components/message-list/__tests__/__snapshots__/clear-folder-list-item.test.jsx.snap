// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearFolderListItem component test suite Selected folder is junk, should render 1`] = `
<div
  className="clearFolderListItem"
>
  <Button
    className="emptyFolder"
    disabled={false}
    icon="delete"
    iconClassName=""
    label="messageList.emptyFolder"
    onClick={[Function]}
    type="button"
  />
  <Translate(ConfirmClearFolderDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</div>
`;

exports[`ClearFolderListItem component test suite Snapshot render Selected folder NOT trash nor junk, should not render 1`] = `""`;

exports[`ClearFolderListItem component test suite Snapshot render Selected folder NOT trash, should not render 1`] = `""`;

exports[`ClearFolderListItem component test suite Snapshot render Selected folder is trash, should render 1`] = `
<div
  className="clearFolderListItem"
>
  <Button
    className="emptyFolder"
    disabled={false}
    icon="delete"
    iconClassName=""
    label="messageList.emptyFolder"
    onClick={[Function]}
    type="button"
  />
  <Translate(ConfirmClearFolderDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</div>
`;
